sched/queue/memory/simple.go

```diff
@@ -3,40 +3,116 @@ package memory
 // In-memory Scheduler Queue
 
 import (
-	"fmt"
 	"github.com/scootdev/scoot/sched"
 	"github.com/scootdev/scoot/sched/queue"
+	"strconv"
 	"time"
 )
 
-func NewSimpleQueue() (queue.Queue, chan queue.WorkItem) {
+func NewSimpleQueue(capacity int) queue.Queue {
 	q := &simpleQueue{}
-	q.outCh = make(chan queue.WorkItem, 1)
-	q.inCh = make(chan enqueueReq)
-	go q.loop()
-	return q, q.outCh
+	q.inCh = make(chan queueReq)
+	q.outCh = make(chan queue.WorkItem)
+
+	st := &simpleQueueState{inCh: q.inCh, outCh: q.outCh, capacity: capacity}
+	go st.loop()
+	return q
 }
 
 type simpleQueue struct {
-	nextID int
-	inCh   chan enqueueReq
-	outCh  chan queue.WorkItem
+	// Immutable state, read by many goroutines
+	inCh  chan queueReq
+	outCh chan queue.WorkItem
+}
+
+type simpleQueueState struct {
+	// Mutable state, only read/written by its goroutine
+	inCh      chan queueReq
+	outCh     chan queue.WorkItem
+	dequeueCh chan struct{}
+	items     []*simpleWorkItem
+	nextID    int
+	capacity  int
 }
 
-type simpleWorkItem sched.Job
+func (s *simpleQueueState) loop() {
+	for !s.done() {
+		s.iter()
 
-func (i simpleWorkItem) Job() sched.Job {
-	return sched.Job(i)
+	}
+	close(s.outCh)
+}
+
+func (s *simpleQueueState) done() bool {
+	// We are done once our inputs are closed and we have no items left to send
+	return s.inCh == nil && len(s.items) == 0
+}
+
+func (s *simpleQueueState) iter() {
+	var outCh chan queue.WorkItem
+	var item *simpleWorkItem
+	// only send if we aren't waiting for an ack and have something to send
+	if s.dequeueCh == nil && len(s.items) > 0 {
+		outCh = s.outCh
+		item = s.items[0]
+	}
+	select {
+	case outCh <- item:
+		s.dequeueCh = item.dequeueCh
+	case req, ok := <-s.inCh:
+		if !ok {
+			s.inCh = nil
+			return
+		}
+		switch req := req.(type) {
+		case enqueueReq:
+			req.resultCh <- s.enqueue(req.def)
+		}
+	case <-s.dequeueCh:
+		// Dequeue the last sent work item
+		s.items = s.items[1:]
+		s.dequeueCh = nil
+	}
+}
+
+func (s *simpleQueueState) enqueue(def sched.JobDefinition) enqueueResult {
+	if len(s.items) >= s.capacity {
+		return enqueueResult{"", queue.NewCanNotScheduleNow(1*time.Second, "queue full")}
+	}
+	id := strconv.Itoa(s.nextID)
+	s.nextID++
+	job := sched.Job{Id: id, Def: def}
+	item := &simpleWorkItem{job, make(chan struct{})}
+	s.items = append(s.items, item)
+	return enqueueResult{id, nil}
 }
 
-func (i simpleWorkItem) Dequeue() {
+type simpleWorkItem struct {
+	job       sched.Job
+	dequeueCh chan struct{}
+}
+
+func (i *simpleWorkItem) Job() sched.Job {
+	return sched.Job(i.job)
```

Comment:
Why cast/construct job, it's already the right type, and you're not doing a deep copy.